src/librustc_mir/transform/nll.rs

```diff
@@ -38,10 +103,20 @@ impl MirPass for NLL {
                           tcx: TyCtxt<'a, 'tcx, 'tcx>,
                           _: MirSource,
                           mir: &mut Mir<'tcx>) {
-        if tcx.sess.opts.debugging_opts.nll {
+        if !tcx.sess.opts.debugging_opts.nll {
+            return;
+        }
+
+        tcx.infer_ctxt().enter(|infcx| {
             // Clone mir so we can mutate it without disturbing the rest
             // of the compiler
-            NLLVisitor::new(tcx).visit_mir(&mut mir.clone());
-        }
+            let mut renumbered_mir = mir.clone();
+
+            // Note that we're using the passed-in mir for the visitor. This is
```

Comment:
Heh, it's either ugly... or genius. I'm not sure which. In any case, if it's only being used for the "origin data" of region variables, we could probably sidestep the question for a bit. For example, instead of using the proper span, we can use dummy-span, and track (for each region variable) the location. Or we can make a `RegionVariableOrigin` that includes the `Location` instead.

I'm sort of inclined to do the former -- just sidestep the existing inference infrastructure, and just try to remember, for each region variable index, what the location was for later. 

Is there a reason that's a bad idea?